packages/protocol/lib/backward/ast-code.ts

```diff
@@ -0,0 +1,380 @@
+      error
+    }
+  }
+}
+
+const doASTCompatibilityReport = (contractName: string, oldAST: ContractAST, newAST: ContractAST): ASTCodeCompatibilityReport => {
+  const oldMethods = createMethodIndex(getCheckableMethodsFromAST(oldAST, 'old'))
+  const newMethods = createMethodIndex(getCheckableMethodsFromAST(newAST, 'new'))
+
+  const report = new ASTCodeCompatibilityReport([])
+
+  // Check for modified or missing methods in the new version
+  Object.keys(oldMethods).forEach((signature: string) => {
+    const method = oldMethods[signature]
+    if (!newMethods.hasOwnProperty(signature)) {
+      // Method deleted, major change
+      report.push(new MethodRemovedChange(contractName, signature))
+      // Continue
+      return
+    }
+    const newMethod = newMethods[signature]
+    report.include(checkMethodCompatibility(contractName, method, newMethod))
+  })
+  // Check for added methods in the new version
+  Object.keys(newMethods).forEach((signature: string) => {
+    if (!oldMethods.hasOwnProperty(signature)) {
+      // New method, minor change
+      report.push(new MethodAddedChange(contractName, signature))
+    }
+  })
+  return report
+}
+
+const generateASTCompatibilityReport = (oldContract: ZContract, oldArtifacts: BuildArtifacts,
+  newContract: ZContract, newArtifacts: BuildArtifacts): ASTCodeCompatibilityReport => {
+  // Sanity checks
+  if (newContract === null) {
+    throw new Error('newContract cannot be null')
+  }
+  if (oldArtifacts === null) {
+    throw new Error('oldArtifacts cannot be null')
+  }
+  if (newArtifacts === null) {
+    throw new Error('newArtifacts cannot be null')
+  }
+  const contractName = newContract.schema.contractName
+
+  // Need to manually use ContractAST since its internal use in ZContract
+  // does not pass the artifacts parameter to the constructor, therefore
+  // forcing a reloading of BuildArtifacts.
+  const newAST = new ContractAST(newContract, newArtifacts)
+  const newKind = newAST.getContractNode().contractKind
+  if (oldContract === null) {
+    if (newKind === CONTRACT_TYPE_CONTRACT) {
+      return new ASTCodeCompatibilityReport([new NewContractChange(contractName)])
+    } else {
+      // New contract added of a non-contract type (library/interface)
+      // therefore no functionality added
+      return new ASTCodeCompatibilityReport([])
+    }
+  }
+
+  // Name sanity check
+  if (oldContract.schema.contractName !== contractName) {
+    throw new Error(`Contract names should be equal: ${oldContract.schema.contractName} !== ${contractName}`)
+  }
+
+  const oldAST = new ContractAST(oldContract, oldArtifacts)
+  const kind = oldAST.getContractNode().contractKind
+  if (kind !== newKind) {
+    // different contract kind (library/interface/contract)
+    return new ASTCodeCompatibilityReport([new ContractTypeChange(contractName, kind, newKind)])
+  }
+
+  const report = doASTCompatibilityReport(contractName, oldAST, newAST)
+  // Check deployed byte code change
```

Comment:
I'll check that out, what that does strip metadata do? The only issue I had so far with metadata is that the deployedBytecode seems to depend on the source folder path, which was annoying to make the test resources compile properly

Reply:
`stripMetadata` erases the part of the bytecode where the metadata hash is stored, so by using it on both bytecodes, we can ensure that bytecode is exactly the same modulo the metadata we don't care about. 

More info here: https://solidity.readthedocs.io/en/v0.5.3/metadata.html#encoding-of-the-metadata-hash-in-the-bytecode
 
Also, just noticed that newer versions of Solidity store this differently. For example 0.5.11 has a compiler version also appended in the metadata, and in e.g. 0.6.10, it's an IPFS hash instead of a swarm hash that's stored. So make sure to take that into account if you end up doing something with this.

Reply:
This seems like it could probably need its own research. What do you think if we leave it for a follow up issue?